src/main/io/gimbal_control.c

```diff
@@ -0,0 +1,224 @@
+    uint64_t crch:8;
+    uint64_t crcl:8;
+}  __attribute__ ((__packed__)) gimbalCalStatus_t;
+
+#define GIMBAL_ROLL_MIN -500
+#define GIMBAL_ROLL_MAX 500
+#define GIMBAL_PITCH_MIN -1150
+#define GIMBAL_PITCH_MAX 1750
+#define GIMBAL_YAW_MIN -2047
+#define GIMBAL_YAW_MAX 2047
+
+static gimbalCmd_t gimbalCmd = {0};
+static serialPort_t *gimbalSerialPort = NULL;
+
+static uint16_t gimbalCrc(uint8_t *buf, uint32_t size)
+{
+    return crc16_ccitt_update(0x0000, buf, size);
+}
+
+// Set the gimbal position on each axis in a ±500 range
+bool gimbalSet(int16_t roll, int16_t pitch, int16_t yaw)
+{
+    DEBUG_SET(DEBUG_GIMBAL, 0, roll);
+    DEBUG_SET(DEBUG_GIMBAL, 1, pitch);
+    DEBUG_SET(DEBUG_GIMBAL, 2, yaw);
+
+    if (!gimbalSerialPort) {
+        return false;
+    }
+
+    // Scale the incoming ±500 range to the max values accepted by the gimbal
+    roll = ((GIMBAL_ROLL_MAX + GIMBAL_ROLL_MIN) / 2) + ((GIMBAL_ROLL_MAX - GIMBAL_ROLL_MIN) * roll / 1000);
+    pitch = ((GIMBAL_PITCH_MAX + GIMBAL_PITCH_MIN) / 2) - ((GIMBAL_PITCH_MAX - GIMBAL_PITCH_MIN) * pitch / 1000);
+    yaw = ((GIMBAL_YAW_MAX + GIMBAL_YAW_MIN) / 2) + ((GIMBAL_YAW_MAX - GIMBAL_YAW_MIN) * yaw / 1000);
+
+    // Constrain to catch any incoming out of range values
+    gimbalCmd.roll = constrain(roll, GIMBAL_ROLL_MIN, GIMBAL_ROLL_MAX);
+    gimbalCmd.pitch = constrain(pitch, GIMBAL_PITCH_MIN, GIMBAL_PITCH_MAX);
+    gimbalCmd.yaw = constrain(yaw, GIMBAL_YAW_MIN, GIMBAL_YAW_MAX);
+
+    DEBUG_SET(DEBUG_GIMBAL, 3, gimbalCmd.roll);
+    DEBUG_SET(DEBUG_GIMBAL, 4, gimbalCmd.pitch);
+    DEBUG_SET(DEBUG_GIMBAL, 5, gimbalCmd.yaw);
+
+    gimbalCmd.mode = gimbalTrackConfig()->gimbal_stabilisation;
+    gimbalCmd.sens = gimbalTrackConfig()->gimbal_sensitivity;
+
+    uint16_t crc = gimbalCrc((uint8_t *)&gimbalCmd, sizeof(gimbalCmd) - 2);
+
+    gimbalCmd.crch = (crc >> 8) & 0xff;
+    gimbalCmd.crcl = crc & 0xff;
+
+    return true;
+}
+
+// Set the gimbal position on each axis in a ±500 range with scale and offset applied
+bool gimbalTrack(int16_t throttle, int16_t roll, int16_t pitch, int16_t yaw)
+{
+    int16_t gimbal_roll;
+    int16_t gimbal_pitch;
+    int16_t gimbal_yaw;
+
+    // Apply scale an offset, each of which are expressed as a percentage
+    gimbal_roll = (roll * gimbalTrackConfig()->gimbal_roll_gain / 100) + 5 * gimbalTrackConfig()->gimbal_roll_offset;
+    gimbal_yaw = (yaw * gimbalTrackConfig()->gimbal_yaw_gain / 100) + 5 * gimbalTrackConfig()->gimbal_yaw_offset;
+
+    if (pitch < 0) {
+        gimbal_pitch = pitch * gimbalTrackConfig()->gimbal_pitch_low_gain / 100;
+    } else {
+        gimbal_pitch = pitch * gimbalTrackConfig()->gimbal_pitch_high_gain / 100;
+    }
+
+    gimbal_pitch += 5 * gimbalTrackConfig()->gimbal_pitch_offset - (throttle * gimbalTrackConfig()->gimbal_pitch_thr_gain / 100);
+
+    return gimbalSet(gimbal_roll, gimbal_pitch, gimbal_yaw);
+}
+
+// Gimbal updates should be sent at 100Hz or the gimbal will self center after approx. 2 seconds
```

Comment:
(isn't 100Hz overkill?)

Reply:
The original API from the vendor states:

> The packet sent by the header to the header will be sent in a continuous cycle [100Hz].

That said it seems to work perfectly well at 10Hz. What should we do here? Next PR will implement a passthrough of the serial data from the VTX, mixing the headtracker input with the above mixing. For that it will probably be best to match the VTX serial update rate.